src/coreclr/src/vm/ceemain.cpp

```diff
@@ -633,6 +633,18 @@ void EEStartupHelper()
     {
         g_fEEInit = true;
 
+#if CORECLR_EMBEDDED
+
+#ifdef TARGET_WINDOWS
+        HINSTANCE curModule = WszGetModuleHandle(NULL);
+#else
+        HINSTANCE curModule = PAL_LoadLibraryDirect(NULL);
```

Comment:
`PAL_LoadLibraryDirect` returns the OS library handle. It is different from the Win32 PAL emulator `HINSTANCE` handle.

Reply:
Are there `g_hmodCoreCLR` and `g_hThisInst` actually used for anything (on Unix)?

Reply:
I saw in other places `WszGetModuleHandle` used in parallel to `PAL_LoadLibraryDirect`, so I assumed they are comparable or otherwise will lead to a quick failure.
You are right though, `HINSTANCE` is the `&exe_module` itself, not the handle.

I now wonder if this is used at all on Unix indeed.

Reply:
I think it is possible it is used and tests do not cover that. It may be safer to use `exe_module`. Trying to figure how to get to it from here.

Reply:
g_hmodCoreCLR is used in the libcoreclr on Unix at two places. First one is in WasThrownByUs, which is called from multiple places.  The second is in CorCompileGetRuntimeDll which is actually unused and can be deleted.
g_hThisInst is used at multiple places even on Unix, but it seems that in those cases might be replaceable by the coreclr path. There are calls to IsIPInModule that use it too, but that function actually always returns false for Unix, so we could ifdef out that call at the caller site. 
But I've just skimmed through the code, so I may have missed something.

Reply:
Perhaps it is safer just to set these to `exe_module`? 
Is there an existing pal api to get that?

Reply:
I could use something like `HINSTANCE curModule = LOADGetPalLibrary()->hinstance;` here, but `LOADGetPalLibrary` is not available outside PAL. 
Perhaps I should make it available as `PAL_GetPalLibrary()`, unless we have another way to get that?

Reply:
None of this seems to be doing anything useful. It would be best to ifdef it out for Windows-only.

Reply:
@jkotas - Are you suggesting just leave them as NULL on Unix?

Reply:
I was a bit concerned that if I just leave it NULL in this change I may miss some obscure cases that use it - directly or indirectly.

It would make sense to do such change for the normal (not linked) case as well and fix everything that breaks.
Maybe makes sense as a separate PR, at least while in finding/fixing phase.

Reply:
Regarding the use of `GetCLRModule()` in `MarkAsThrownByUsWorker` and `WasThrownByUs` - any idea what can be used instead?
It seems like any other "unique" value could be used as long as it is consistent between these uses, perhaps even `1`.

Reply:
On Unix, MarkAsThrownByUsWorker can be no-op and WasThrownByUs can always return true.

Reply:
> g_hThisInst is used at multiple places even on Unix, but it seems that in those cases might be replaceable by the coreclr path

What is the conventional way of getting coreclr path?
Typical use is like:
```c
            if (!WszGetModuleFileName(GetCLRModule(), dllPath)) {
                dllPath.Set(W("\0"));
            }
```
`g_hThisInst` is also used for this purpose.

Reply:
Actually, it seems we can just change GetCLRModule to return GetModuleHandle(NULL) for single file host. I was thinking that returning NULL would work too, as the WszGetModuleFileName would still return the current exe path. But we need it to be a valid module handle on Windows in the MarkAsThrownByUsWorker / WasThrownByUs. Unless we change that method to use something else for tagging, which I don't think is worth it.

Reply:
I do not think spreading the differences between normal (.dll) and single-file (.exe) to the other parts of coreclr is desirable. If we are going to make `g_hmodCoreCLR` and `g_hThisInst` not available on Unix, it should be not available regardless dll or exe.

I have tried to experiment with such change in this PR - https://github.com/dotnet/runtime/pull/43735

I have ifdef'd out only `g_hmodCoreCLR` for starter, since it seems easier of the two.
Let me know if this is not going in the direction that you've expected.

It looks like it can be done, but there are still 2 issues:

- GetCLRModule() does not seem to be completely useless in `WasThrownByUs`. Just returning `TRUE` results in lots of test failures. Returning `FALSE` also results in failures. I guess both `TRUE` and `FALSE` happen, even on Linux.
I did not look deeper at what `GetCLRModule()` is used for. On the first glance it looks like it is used just as a non-zero flag. Indeed using `42` instead seems to work as well.
- `g_hThisInst` is used to get the path to the containing module (coreclr.dll or whatever.exe). I am using "\0" for now.  
We would need some alternative API to get to the path. 
There are more cases where `g_hThisInst` is used for that purpose as well.

So far it still feels like setting `g_hThisInst` and `g_hmodCoreCLR` to be `&exe_module` in .exe scenario and leaving the rest as-is and consistent with ".dll" case might be a cleaner solution to all this.

Reply:
> * I am using "\0" for now.

That doesn't seem correct. It would return path to the exe even in non-single file mode while we want the coreclr.dll path.

Reply:
It is literally "" in the experiment just to get it unblocked. 

Ideally I would call an API that returns the path to the coreclr module. Before the change it was `WszGetModuleFileName(GetCLRModule(), dllPath)`, which worked for either .dll or .exe case.
If the module handle is unavailable, we need something else on Unix. 
I have not yet figured what I could call instead, so I just return "" for now in the experiment.

Reply:
Ah, I've misread the "\0", I have thought you meant using NULL.

Reply:
Having these difficulties make me think that the module handle itself is not useless and its uses are not broken.

The part that I set it incorrectly in this PR is a bug - no questions about that. 
But probably it is the only part that needs fixing.

Reply:
PAL  contains a pal_module global variable and LOADGetPalLibrary function to get it. That should work for GetCLRModule() if you exposed it for both coreclr.dll and single file host cases.
There is also a PAL_GetPALDirectoryW that could get used instead of the WszGetModuleFileName which seems to always be followed by cutting off the filename and caring for the path only.

Reply:
Yes, `LOADGetPalLibrary ` is the thing. 
It looks like if I store result of `LOADGetPalLibrary` into `g_hmodCoreCLR` and `g_hThisInst` in a single file case, every use will just work.

That is attractive because then only `EEStartupHelper` will need to know about statically linked case and it needs to know anyways.

Reply:
I have changed this line to be `HINSTANCE curModule = PAL_GetPalHostModule();` on Unix.

Let me know if we still want to proceed with removal of `g_hmodCoreCLR` and `g_hThisInst`, although I think it is better if we leave them as-is. 
FWIW, they are used on Unix. It is possible to remove them but it would require changes, like switching to something else in a few places. It would not be a simple removal or substitution and it is not directly related to how coreclr/jit are linked, so it feels better to do that in a separate PR.